### ISSUE TITLE
conv~ partition scheme

### ISSUE THREAD

**Issue by porres** [porres/pd-else#1859]

**Comment by porres** [comment]
@brbrofsvl we now got a compiled convolution object based on brent's convolve

I remember you wanted to work on a convolution object with a new partition scheme, right?

also pinging @timothyschoen into the discussion.

It'd be great if we could offer something new and different

**Comment by timothyschoen** [comment]
> @brbrofsvl we now got a compiled convolution object based on brent's convolve
> 
> I remember you wanted to work on a convolution object with a new partition scheme, right?
> 
> also pinging @timothyschoen into the discussion.
> 
> It'd be great if we could offer something new and different

Tbh, I don't think there is much value in using a different partitioning scheme just for the sake of it? If the current one works and is performant enough, might as well stick with that?

**Comment by porres** [comment]
> vTbh, I don't think there is much value in using a different partitioning scheme just for the sake of it? If the current one works and is performant enough, might as well stick with that?

there are benefits for the gardner scheme. It is more efficient. In fact, see https://lists.puredata.info/pipermail/pd-list/2019-01/124278.html where you can see that the abstraction "pp.fft-partconv~" from audiolab is more efficient than brent's external. That's right, an abstraction more efficient than an external, so how better would it be if it were a compiled external, huh?

Now, this is mostly useful for a 0 latency, which is nice, and would also mean a 64 partition size. Using a big IR table with 64 size is very very hard on my old machine, an example is the 'church' IR from my help file.

So, an efficient 0 delay external would be KILLER :)

**Comment by timothyschoen** [comment]
It's actually a lot less efficient, though. The paper only claims it to be faster than direct form FIRs, which makes sense because direct form FIRs are very slow. It's slower than large FFT based FIRs, which is what we do.

<img width="639" alt="Screenshot 2023-09-21 at 15 25 13" src="https://github.com/porres/pd-else/assets/44585538/6b610548-bd22-446a-a7de-e1b285121005">


It does reduce output delay, so that could still be a reason to do it.

**Comment by brbrofsvl** [comment]
There are better schemes than the Gardner one. The reason larger blocks work so well it's that it greatly reduces computation in the overlap-add stage. Zero delay is desirable for a general FIR convolution filter, including reverb, so a compromise is necessary, trading a few more FFTs to reduce overlap-adds. (I'm writing this for the benefit of anyone who might come across it later). There are lots of other things that can be done, like attempting to stagger FFTs so they don't all come due at exactly the same tick.

The strategy for a compiled object could be different from that for a Pd patch, where in the latter just reducing the overall number of objects can be a huge gain. 

What's the timeline for this?

**Comment by porres** [comment]
> It's actually a lot less efficient, though. 

but @timothyschoen this is not comparing to a partitioned convolution of size 64

**Comment by porres** [comment]
> There are better schemes than the Gardner one. 

like what @brbrofsvl ? and how is it different? If I'm not mistaken, Gardner starts with a minmum size and keeps increasing, right?

Object is here already, but I'd hope we could change and work on it before the next update, like in a few weeks or so. I'd like to hold until this is ready anyway... you people can also give me hints so I can try and work on it myself with a little bit of your help

**Comment by brbrofsvl** [comment]
Yeah, that's what happens with Gardner. It typically grows as fast as possible. The trouble is that a new FFT can be as expensive as a few more blocks of overlap-add, so it can be better to reduce the total number of FFTs. There are papers that show an algorithm for optimizing, but it's hard to be sure on different architectures. You can go to like 6.4 or 6.5 in this paper for a review:

https://publications.rwth-aachen.de/record/466561/files/466561.pdf
